### PR TITLE
Upgrade to SDK 48 and Detox 20

### PR DESCRIPTION
**Pull request by douglowder**
- Move to SDK 48
- Move to Detox 20
- Update Detox configuration files and `openApp.js` for Detox 20
- Improve EAS on success hook
- Invoke screenshot on failure from command line (fix issue #7 )
- Should be merged at the same time as docs PR https://github.com/expo/expo/pull/21753

### CODE REVIEW

File: e2e/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/** @type {import('jest').Config} */
+module.exports = {
+  setupFilesAfterEnv: ['./setup.js'],
+  reporters: ['detox/runners/jest/reporter'],
+  globalSetup: 'detox/runners/jest/globalSetup',
+  globalTeardown: 'detox/runners/jest/globalTeardown',
+  testEnvironment: 'detox/runners/jest/testEnvironment',
+  testTimeout: 120000,
+  testMatch: ['./**/*.e2e.js'],
```
Comment (dsokal) [inline e2e/jest.config.js]: Can we use the default here? 
```suggestion
  rootDir: '..',
  testMatch: **['<rootDir>/e2e/**/*.e2e.js'],**
```

File: detox.config.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/** @type {Detox.DetoxConfig} */
+module.exports = {
+  logger: {
+    level: process.env.CI ? 'debug' : undefined,
+  },
```
Comment (dsokal) [inline detox.config.js:5]: What's the behavior for a non-CI environment? Does it default to `info`? 
I can see Detox docs have this line written slightly different - https://wix.github.io/Detox/docs/config/overview

File: detox.config.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  testRunner: {
+    args: {
+      config: 'e2e/jest.config.js',
+      maxWorkers: 1,
+    },
+  },
+  artifacts: {
+    plugins: {
+      log: process.env.CI ? 'failing' : undefined,
+      screenshot: process.env.CI ? 'failing' : undefined,
```
Comment (dsokal) [inline detox.config.js]: Don't we want to take screenshots when tests fail locally? This worked differently before.

File: detox.config.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    },
+  },
+  apps: {
+    'ios.debug': {
+      type: 'ios.app',
+      binaryPath: 'ios/build/Build/Products/Debug-iphonesimulator/BareExpoDetox.app',
+      build: './scripts/build-detox-ios.sh Debug',
```
Comment (dsokal) [inline detox.config.js]: This script seems to be missing in this repository. Did you forget to commit it? 🤔

File: eas-hooks/eas-build-on-success.sh [hunk 5]
```diff
@@ -17,20 +33,23 @@ if [[ "$EAS_BUILD_PLATFORM" == "android" ]]; then
...
-  adb emu kill
 else
+  # Execute iOS tests
   if [[  "$EAS_BUILD_PROFILE" == "test" ]]; then
-    detox test --configuration ios.release --headless
+    detox test --configuration ios.release --take-screenshots=failing
   fi
   if [[ "$EAS_BUILD_PROFILE" == "test_debug" ]]; then
-    detox test --configuration ios.debug --headless
+    detox test --configuration ios.debug --take-screenshots=failing
```
Comment (dsokal) [inline eas-hooks/eas-build-on-success.sh]: Is it necessary to pass `--take-screenshots=failing` when it's already configured here - https://github.com/expo/eas-tests-example/pull/8/files#diff-c199d282e0ab6b6dcfee2219852b5d5f2e9a2f3233dcda912eb7bd55d264a42eR15 ?

File: eas-hooks/eas-build-on-success.sh [hunk 6]
```diff
@@ -17,20 +33,23 @@ if [[ "$EAS_BUILD_PLATFORM" == "android" ]]; then
     counter=$((counter + 1))
   done
 
+  sleep 10
```
Comment (dsokal) [inline eas-hooks/eas-build-on-success.sh]: What problem is this sleep solving? Every time I see sth like this I immediately think this is some kind of hack. I think this requires at least a comment.

**Review by dsokal — COMMENTED** [review]
Generally, this looks really great! Thanks for taking care of this.
Please address/answer a couple of comments before I approve this PR.

--- 

I think we also need the docs PR before we can merge this one.

File: detox.config.js (hunk 3 above)
Comment (douglowder) [reply detox.config.js]: Agreed this should always be "failing"

File: detox.config.js (hunk 4 above)
Comment (douglowder) [reply detox.config.js]: Not sure this script is needed, will check this

File: eas-hooks/eas-build-on-success.sh (hunk 5 above)
Comment (douglowder) [reply eas-hooks/eas-build-on-success.sh]: Agreed, the config should automatically be enabling the screenshots

File: eas-hooks/eas-build-on-success.sh (hunk 6 above)
Comment (douglowder) [reply eas-hooks/eas-build-on-success.sh]: This is what we are doing in updates E2E to help ensure that the Android emulator is ready. Probably not needed here, agreed.

File: detox.config.js (hunk 3 above)
Comment (douglowder) [reply detox.config.js]:
```suggestion
      screenshot: 'failing',
```

File: detox.config.js (hunk 4 above)
Comment (douglowder) [reply detox.config.js]:
```suggestion
```

File: eas-hooks/eas-build-on-success.sh (hunk 6 above)
Comment (douglowder) [reply eas-hooks/eas-build-on-success.sh]:
```suggestion
```

File: detox.config.js (hunk 2 above)
Comment (douglowder) [reply detox.config.js:5]: This appears to be the same as the Detox example app: https://github.com/wix/Detox/blob/master/examples/demo-react-native/detox.config.js

**Comment by douglowder** [comment]
@dsokal I think I've addressed all the issues -- I'll work on the associated docs PR before coming back to this.

**Review by dsokal — APPROVED** [review]
Looks great, thanks!

**Pull request merged**